File: src/tune_block_sizes.cpp
#include <iostream>
#include <limits>
#include <vector>

#include <stream_compaction/efficient.h>
#include <stream_compaction/naive.h>

#include "testing_helpers.hpp"

template <typename Timer, typename Func>
float measureTime(Timer &timer, Func func) {
    const int warmup = 10;
    const int iterations = 100;
    for (int i = 0; i < warmup; i++) {
        func();
    }
    float totalTime = 0.0f;
    for (int i = 0; i < iterations; ++i) {
        func();
        totalTime += timer.getGpuElapsedTimeForPreviousOperation();
    }
    return totalTime / iterations;
}

template <typename Timer, typename Func, typename... Args>
void tuneBlockSize(Timer &timer, Func func, Args &&...args) {
    std::vector<int> blockSizes = {8, 16, 32, 64, 128, 256, 512, 1024};
    float bestTime = std::numeric_limits<float>::max();
    int bestBlockSize = -1;
    for (int blockSize : blockSizes) {
        float time = measureTime(timer, [&]() { func(std::forward<Args>(args)..., blockSize); });
        std::cout << "Block size: " << blockSize << ", time: " << time << " ms" << std::endl;
        if (time < bestTime) {
            bestTime = time;
            bestBlockSize = blockSize;
        }
    }
    std::cout << "Best block size: " << bestBlockSize << std::endl;
}

int main() {
    const int SIZE = 1 << 22;
    const int NPOT = SIZE - 3;
    int *a = new int[SIZE];
    int *b = new int[SIZE];
    int *c = new int[SIZE];

    genArray(SIZE - 1, a, 50);
    a[SIZE - 1] = 0;

    zeroArray(SIZE, c);
    printDesc("naive scan, power-of-two");
    tuneBlockSize(StreamCompaction::Naive::timer(), StreamCompaction::Naive::scan, SIZE, c, a);

    zeroArray(SIZE, c);
    printDesc("naive scan, non-power-of-two");
    tuneBlockSize(StreamCompaction::Naive::timer(), StreamCompaction::Naive::scan, NPOT, c, a);

    zeroArray(SIZE, c);
    printDesc("work-efficient scan, power-of-two");
    tuneBlockSize(StreamCompaction::Efficient::timer(), StreamCompaction::Efficient::scan, SIZE, c,
                  a);

    zeroArray(SIZE, c);
    printDesc("work-efficient scan, non-power-of-two");
    tuneBlockSize(StreamCompaction::Efficient::timer(), StreamCompaction::Efficient::scan, NPOT, c,
                  a);

    genArray(SIZE - 1, a, 4);
    a[SIZE - 1] = 0;

    zeroArray(SIZE, c);
    printDesc("work-efficient compact, power-of-two");
    tuneBlockSize(StreamCompaction::Efficient::timer(), StreamCompaction::Efficient::compact, SIZE,
                  c, a);

    zeroArray(SIZE, c);
    printDesc("work-efficient compact, non-power-of-two");
    tuneBlockSize(StreamCompaction::Efficient::timer(), StreamCompaction::Efficient::compact, NPOT,
                  c, a);

    system("pause");
    delete[] a;
    delete[] b;
    delete[] c;

    return 0;
}

File: scripts/tune_block_sizes.py
from helper import measure_time


def tune_block_size(operation: str, implementation: str, input_size: int) -> None:
    block_sizes = [8, 16, 32, 64, 128, 256, 512, 1024]
    best_time = float("inf")
    best_block_size = -1
    for block_size in block_sizes:
        time_sum = 0.0
        time_count = 10
        for _ in range(time_count):
            time_sum += measure_time(operation, implementation, input_size, block_size)
        time = time_sum / time_count
        print(f"Block size: {block_size}, time: {time} ms")
        if time < best_time:
            best_time = time
            best_block_size = block_size
    print(f"Optimal block size: {best_block_size}, time: {best_time} ms")


def print_divider() -> None:
    print("=" * 40)


def main():
    input_size = 1 << 22

    print("Naive scan, power-of-two")
    tune_block_size("scan", "naive", input_size)
    print_divider()
    print("Naive scan, non-power-of-two")
    tune_block_size("scan", "naive", input_size - 3)
    print_divider()
    print("Work-efficient scan, power-of-two")
    tune_block_size("scan", "efficient", input_size)
    print_divider()
    print("Work-efficient scan, non-power-of-two")
    tune_block_size("scan", "efficient", input_size - 3)
    print_divider()
    print("Work-efficient compact, power-of-two")
    tune_block_size("compact", "efficient", input_size)
    print_divider()
    print("Work-efficient compact, non-power-of-two")
    tune_block_size("compact", "efficient", input_size - 3)


if __name__ == "__main__":
    main()
